src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -389,6 +554,32 @@ ParsedQuery Visitor::visitTypesafe(Parser::SelectQueryContext* ctx) {
                        ctx->getText());
 }
 
+// ____________________________________________________________________________________
+GraphPatternOperation::Subquery Visitor::visitTypesafe(
+    Parser::SubSelectContext* ctx) {
+  ParsedQuery query;
+  visibleVariables_.emplace_back();
+  query._clause = visitTypesafe(ctx->selectClause());
+  query._rootGraphPattern = visitTypesafe(ctx->whereClause());
+  query.setNumInternalVariables(numInternalVariables_);
+  query.addSolutionModifiers(visitTypesafe(ctx->solutionModifier()));
+  numInternalVariables_ = query.getNumInternalVariables();
+  auto values = visitTypesafe(ctx->valuesClause());
+  if (values.has_value()) {
+    query._rootGraphPattern._children.emplace_back(std::move(values.value()));
+  }
```

Comment:
Are you sure that this is the correct semantics for this `values` clause at the end of the query?
You are treating it as an ordinary `inlineData`....

Reply:
Like I read it, it is rather a values clause in the OUTER query... So we have to do some more research I think.

Reply:
But then again, it is still inside the same braces as the subquery... we really have to check what implications this has.

Reply:
I had a look at [10.2](https://www.w3.org/TR/sparql11-query/#inline-data) of the standard. I think you're right. The old code should be wrong. As I understand it the Values apply on the next outer GraphPattern of the SubSelect. So they apply on the variables selected by the SubSelect and not the Variables inside the SubSelects WhereClause. This makes a difference because not all variables are exposed but also because they could simply be renamed in the Select.
But because the SubSelect will always be in another GraphPattern this should still be implementable.

Reply:
Yes, thanks for finding out the section. I read this the same way. I think the easiest way to implement this is to additionally return the `optional<Values>` from the `visitSubselect` and then add it in the calling code. I think it is feasible to directly integrate this here.

Reply:
Yes that sounds like what I implemented in the mean time :)